### PR TITLE
chore: refactor for clarity

### PR DESCRIPTION
**Pull request by codyzu**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Refactored implementation for readability.

These important changes were made:
1. I updated the readme to reflect the actual behavior for the value returned when the `origin` option is a function. The docs simply stated it should return `true` or `false` to allow the origin, but the implementation actually use the return value to process to the origin (i.e. it could be a string). This aligns fastify-cors [express-cors](https://github.com/expressjs/cors#configuration-options).
2. If the `origin` option is a function and returns a falsy value other than `false`, for example `undefined`, the hook generates an error. The previous implementation treated this as `*` origin which does not seem intuitive and is not the way that express-cors works.
3. If the `origin` option is a function and throws an error, the error is caught and passed into the hook callback (generating an error).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,15 +35,16 @@ You can use it as is without passing any option, or you can configure it as expl
...
-  - `Function` - set `origin` to a function implementing some custom logic. The function takes the request origin as the first parameter and a callback as a second (which expects the signature `err [Error | null], allow [bool]`), *async-await* and promises are supported as well. Fastify instance is bound to function call and you may access via `this`. For example: 
+  - `Function` - set `origin` to a function implementing some custom logic. The function takes the request origin as the first parameter and a callback as a second (which expects the signature `err [Error | null], origin`), where `origin` is a non function value of the origin option. *Async-await* and promises are supported as well. The Fastify instance is bound to function call and you may access via `this`. For example: 
   ```js
   origin: (origin, cb) => {
     if(/localhost/.test(origin)){
       //  Request from localhost will pass
       cb(null, true)
       return
     }
-    cb(new Error("Not allowed"), false)
```
Comment (codyzu) [inline README.md:46]: This example seems unnecessarily complicated... calling with an error *and* false.

File: index.js [hunk 2]
```diff
@@ -110,66 +78,102 @@ function fastifyCors (fastify, opts, next) {
...
+  async function resolveOriginWrapper (req, cb) {
+    try {
+      const result = origin.call(fastify, req.headers.origin, cb)
+
+      // Allow for promises
+      if (result && typeof result.then === 'function') {
+        try {
+          return cb(null, await result)
+        } catch (error) {
+          return cb(error)
```
Comment (codyzu) [inline index.js:138]: catch async errors

File: index.js [hunk 3]
```diff
@@ -110,66 +78,102 @@ function fastifyCors (fastify, opts, next) {
...
       }
-      return false
-    } else if (typeof origin === 'string') {
-      return reqOrigin === origin
-    } else if (origin instanceof RegExp) {
-      return origin.test(reqOrigin)
-    } else {
-      return !!origin
+    } catch (error) {
+      cb(error)
```
Comment (codyzu) [inline index.js:142]: catch sync errors

File: index.js [hunk 4]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
     strictPreflight: true
   }, opts)
 
-  const isOriginFalsy = !origin
-  const isOriginString = typeof origin === 'string'
-  const isOriginFunction = typeof origin === 'function'
+  const resolveOriginOption = typeof origin === 'function' ? resolveOriginWrapper : (_, cb) => cb(null, origin)
```
Comment (mcollina) [inline index.js:32]: I don't see where this is called.

File: index.js [hunk 5]
```diff
@@ -110,66 +78,102 @@ function fastifyCors (fastify, opts, next) {
...
+      .header('Content-Length', '0')
+      .send()
   }
 
-  function isOriginAllowed (reqOrigin, origin) {
-    if (Array.isArray(origin)) {
-      for (let i = 0; i < origin.length; ++i) {
-        if (isOriginAllowed(reqOrigin, origin[i])) {
-          return true
+  async function resolveOriginWrapper (req, cb) {
```
Comment (mcollina) [inline index.js:129]: This is _really bad_ and it can cause huge amount of problems. Never mix callbacks callbacks and async.

File: index.js (hunk 2 above)
Comment (mcollina) [reply index.js:138]: don't nest try-catch

File: index.js (hunk 3 above)
Comment (mcollina) [reply index.js:142]: Unfortunately this can lead to `cb` being called twice, and possibly generating an unhandledRejection.

File: test/cors.test.js [hunk 6]
```diff
@@ -252,10 +288,6 @@ test('Dynamic origin resolution (errored - promises)', t => {
   }
   fastify.register(cors, { origin })
 
-  fastify.get('/', (req, reply) => {
-    reply.send('ok')
-  })
```
Comment (mcollina) [inline test/cors.test.js:257]: why was this deleted?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js [hunk 7]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
...
   function onRequest (req, reply, next) {
     // Always set Vary header
     // https://github.com/rs/cors/issues/10
     vary(reply, 'Origin')
 
-    if (isOriginFalsy) {
-      req.corsOriginAllowed = false
-      return next()
-    }
+    resolveOriginOption(req, (error, resolvedOriginOption) => {
```
Comment (codyzu) [inline index.js:48]: @mcollina it's called here

File: index.js (hunk 4 above)
Comment (codyzu) [reply index.js:32]: line 48

File: test/cors.test.js (hunk 6 above)
Comment (codyzu) [reply test/cors.test.js:257]: It was never called. The test results in an error in the onRequest hook and the route never gets called. I removed it to try to make the tests consistent (remove unnecessary routes).

**Comment by codyzu** [comment]
_* All tests reporting `97.5% req / sec` results._

| test | run 1 | run 2 | run 3 |
|--|--|--|--|
| before normal request `GET` | 33311 | 33471 | 33183 |
| before preflight request `OPTIONS` | 37727 | 37343 | 37375 |
| after normal request `GET` | 34687 | 33503 | 32927 |
| after preflight request `OPTIONS` | 36863 | 37247 | 36767 |

@mcollina Here are my performance indicators comparing the previous implementation (before) to my current changes (after). Big picture, the change seems relatively small. Interestingly, the test results before my changes have less variation between test runs. Not sure what causes that... might just be a coincidence related to the load on my machine :woman_shrugging:

**Review by mcollina — APPROVED** [review]
lgtm

File: index.js [hunk 8]
```diff
@@ -110,66 +78,94 @@ function fastifyCors (fastify, opts, next) {
         )
       }
 
-      next()
+      return next()
```
Comment (Ethan-Arrowood) [inline index.js:81]: Why are we returning the next call?

Comment (codyzu) [reply index.js:81]: that is probably an artifact from my refactor... I can remove it :smile:

Comment (codyzu) [reply index.js:81]: Actually, that makes this return consistent with all of the other return paths. Does it matter? In any case, I would propose that we are consistent. I'll note the other places where next is returned...

File: index.js [hunk 9]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
...
+      // Falsy values are invalid
+      if (!resolvedOriginOption) {
+        return next(new Error('Invalid CORS origin option'))
+      }
 
-      req.corsOriginAllowed = origin
+      // Enable preflight
+      req.corsOriginAllowed = true
 
-      if (origin === false) return next()
```
Comment (codyzu) [inline index.js:100]: @Ethan-Arrowood here returning the result of next

File: index.js [hunk 10]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
...
+      }
+
+      // Disable CORS and preflight if false
+      if (resolvedOriginOption === false) {
+        req.corsOriginAllowed = false
+        return next()
+      }
 
-    configureOrigin(req, reply, (err, origin) => {
-      if (err !== null) return next(err)
```
Comment (codyzu) [inline index.js:96]: @Ethan-Arrowood here returning the result of next

File: index.js [hunk 11]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
...
 
-  fastify.addHook('onRequest', onRequest)
   function onRequest (req, reply, next) {
     // Always set Vary header
     // https://github.com/rs/cors/issues/10
     vary(reply, 'Origin')
 
-    if (isOriginFalsy) {
-      req.corsOriginAllowed = false
-      return next()
```
Comment (codyzu) [inline index.js:92]: @Ethan-Arrowood here returning the result of next

File: index.js (hunk 8 above)
Comment (codyzu) [reply index.js:81]: I don't have an opinion, I propose just that we use the same pattern everywhere. Let me know what is best...

Comment (Ethan-Arrowood) [reply index.js:81]: Good point! I reviewed this on mobile so it was tough to see how other lines were doing it.

**Review by Ethan-Arrowood — APPROVED** [review]
Lgtm

File: index.js [hunk 12]
```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
...
+      if (resolvedOriginOption === false) {
+        req.corsOriginAllowed = false
+        return next()
+      }
 
-    configureOrigin(req, reply, (err, origin) => {
-      if (err !== null) return next(err)
+      // Falsy values are invalid
+      if (!resolvedOriginOption) {
+        return next(new Error('Invalid CORS origin option'))
```
Comment (RafaelGSS) [inline index.js:61]: This is new behavior? In past code, we are setting `req.corsOriginAllowed = false` on falsy values.

Comment (RafaelGSS) [reply index.js:61]: If is an intentional behavior, should be tested as well.

File: test/cors.test.js [hunk 13]
```diff
@@ -160,9 +160,25 @@ test('Dynamic origin resolution (errored)', t => {
...
   })
+})
+
+test('Dynamic origin resolution (invalid result)', t => {
+  t.plan(3)
+
+  const fastify = Fastify()
+  const origin = (header, cb) => {
+    t.strictEqual(header, 'example.com')
+    cb(null, undefined)
```
Comment (codyzu) [inline test/cors.test.js:179]: @RafaelGSS this test covers falsy values. The test is in the context of an origin function that returns a falsy value, but covers that case. Is ok for you?

File: index.js (hunk 12 above)
Comment (codyzu) [reply index.js:61]: You are correct that this is a functional change. Before, if the origin option was a falsy value, CORS would just be disabled. If the origin option was a function that returned a falsy value, the accept origin header would be `*`.

This was an attempt to make the behavior from the origin option predictable in the following ways:
1. The same behavior if the origin option is a value OR a function 
2. If the option is an invalid value (defined as any value not defined in the documentation), an error is thrown. 

Clearly there is some risk to a change like this. However I would argue that it will actual make the module more understandable because the behavior is consistent. I welcome any discussion on this point 😄

File: index.js (hunk 8 above)
Comment (codyzu) [reply index.js:81]: No worries, I’m glad to see you noticed the change to an unassuming line like this 😉

File: index.js [hunk 14]
```diff
@@ -110,66 +78,94 @@ function fastifyCors (fastify, opts, next) {
...
-    if (isOriginFunction) {
-      const result = origin.call(fastify, reqOrigin, _onOrigin)
-      if (result && typeof result.then === 'function') {
-        result.then(res => _onOrigin(null, res), callback)
-      }
-    } else {
-      _configureOrigin(origin)
+  function preflightHandler (req, reply) {
+    // Do not handle preflight requests if the origin was not allowed
+    if (!req.corsOriginAllowed) {
```
Comment (zekth) [inline index.js:87]: This is a new behaviour. Also technically the `404` code is not fitting the spec as the ressource is here but the CORS is not allowed. I'd rather use `400`. Returning a `404` would be handled in a custom function.

Comment (codyzu) [reply index.js:87]: Glad to see an observant eye. This was changed in my last PR #96 . 

This is due to the fact that I moved the preflight logic into the `OPTIONS *` route handler. There was some certain motivation for that (discussed in #96 ) but the general points were to:
1. allow overriding of `OPTIONS` routes (seems to be the motivation behind #88 )
1. make fastify-cors fit better into the fastify plugin paradigm. The previous implementation seemed to be a fairly strict port of express-cors, where middleware can either handle a request or pass it long to the next middleware in the chain. That model doesn't fit with plugins, hooks, and a "static" routing table that fasitfy uses.

Back to the question. This line is to allow for the case when the `origin` option is `false` or is a function that returns `false`. According to the docs, this should disable CORS (and the preflight handler). With middleware, this can just be checked when the middleware executes. With a static route, the route will always exist. 404 is to "simulate" the route not being there.

From a client perspective, if CORS is disabled, shouldn't the `OPTIONS` routes added by fastify-cors be "not found"? 

Let me know if I didn't understand your point. Maybe the name of the field is misleading... perhaps it should be `corsEnabled` similar???

File: index.js (hunk 12 above)
Comment (jsumners) [reply index.js:61]: If your goal is to "refactor for clarity" then you should not also be introducing new functionality. Doing so makes the PR way more difficult to evaluate.

Comment (RafaelGSS) [reply index.js:61]: I agree with @jsumners. I would like to see this change, but it's better in another PR.

Comment (codyzu) [reply index.js:61]: I agree with your point.

The challenge is that the old behavior was a bit tricky (and possibly unintentional):
* if the `origin` option is falsy, CORS is disabled :ballot_box_with_check: 
* if the `origin` option is a function that returns `false`, CORS is disabled :ballot_box_with_check: 
* if the `origin` option is a function that returns a falsy value **other than** `false`, the allowed origin is `*`. (Note this behavior was never documented or tested) :scream: 

:point_up: Handling the last case will add a lot of noise, but should be possible... I'll see if I can separate it into 2 PRs...

Comment (jsumners) [reply index.js:61]: It may not be ideal to replicate weird behavior. But consider that you have reviewers who probably know less about the code than you do at this point and they need to read the original code to understand the new code in order to properly review it. Any functionality change will be a radical change in this respect, thus making it prohibitively more difficult to review. For this case, I would note the change I want to make and then submit a follow up PR after the refactor PR is accepted.

Comment (codyzu) [reply index.js]: @RafaelGSS @jsumners I removed this change. I'll put it into another PR.

File: index.js (hunk 14 above)
Comment (codyzu) [reply index.js]: I renamed this to `corsPreflightEnabled` which better describes the flag.

**Comment by codyzu** [comment]
I removed any possible functional changes. This included:
1. throwing an error if the origin option is a falsy non-false value
1. disabling the preflight logic when the origin option _only_ resolves to false (it will now be disabled for any falsy value, as was the case before).

**Review by jsumners — APPROVED** [review]
It's been so long since I have implemented the CORS spec, I can't really speak to the implementation. But the code looks sane to me.

**Review by RafaelGSS — APPROVED** [review]
LGTM. Good work!

**Comment by mcollina** [comment]
@zekth you approve?

**Review by zekth — APPROVED** [review]
LGTM . Thanks for making it clearer
